server/src/main/java/org/opensearch/search/backpressure/SearchBackpressureManager.java

```diff
@@ -0,0 +1,250 @@
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.backpressure;
+
+import com.sun.management.OperatingSystemMXBean;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.search.SearchShardTask;
+import org.opensearch.common.Streak;
+import org.opensearch.common.SuppressForbidden;
+import org.opensearch.common.util.TokenBucket;
+import org.opensearch.monitor.jvm.JvmStats;
+import org.opensearch.search.backpressure.trackers.ResourceUsageTracker;
+import org.opensearch.tasks.CancellableTask;
+import org.opensearch.tasks.Task;
+import org.opensearch.tasks.TaskResourceTrackingService;
+import org.opensearch.tasks.TaskResourceTrackingService.TaskCompletionListener;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.lang.management.ManagementFactory;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.DoubleSupplier;
+import java.util.function.LongSupplier;
+import java.util.stream.Collectors;
+
+/**
+ * SearchBackpressureManager is responsible for monitoring and cancelling in-flight search tasks if they are
+ * breaching resource usage limits when the node is in duress.
+ */
+@SuppressForbidden(reason = "OperatingSystemMXBean#getProcessCpuLoad")
+public class SearchBackpressureManager implements Runnable, TaskCompletionListener {
+    private static final Logger logger = LogManager.getLogger(SearchBackpressureManager.class);
+    private static final OperatingSystemMXBean osMXBean = (OperatingSystemMXBean) ManagementFactory.getOperatingSystemMXBean();
+
+    private final SearchBackpressureSettings settings;
+    private final TaskResourceTrackingService taskResourceTrackingService;
+    private final List<ResourceUsageTracker> trackers;
+
+    private final Streak cpuBreachesStreak = new Streak();
+    private final Streak heapBreachesStreak = new Streak();
+
+    private final AtomicLong completionCount = new AtomicLong();
+    private final AtomicLong cancellationCount = new AtomicLong();
+    private final AtomicLong limitReachedCount = new AtomicLong();
+
+    private final TokenBucket taskCancellationRateLimiter;
+    private final TokenBucket taskCancellationRatioLimiter;
+
+    private final LongSupplier timeNanosSupplier;
+    private final DoubleSupplier cpuUsageSupplier;
+    private final DoubleSupplier heapUsageSupplier;
+
+    public SearchBackpressureManager(
+        SearchBackpressureSettings settings,
+        TaskResourceTrackingService taskResourceTrackingService,
+        ThreadPool threadPool
+    ) {
+        this(
+            settings,
+            taskResourceTrackingService,
+            threadPool,
+            System::nanoTime,
+            osMXBean::getProcessCpuLoad,
```

Comment:
Can we simply use ProcessProbe#getProcessCpuPercent instead of taking a direct dependency on `OperatingSystemMXBean `

Reply:
Acknowledged; will update the CPU usage supplier.